NATIVE-306: News stream feature

amusameh opened the pull request:
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/). 
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **WEBAPP**.

mawadkittos commented on src/modules/app/PayloadDataType.js:
```diff
@@ -6,8 +6,10 @@ import {
   LanguageModel,
-  PageModel, SprungbrettJobModel, WohnenOfferModel
+  PageModel, SprungbrettJobModel, WohnenOfferModel,
+  LocalNewsModel, TunewsModel
```
@amusameh I think we should change to TuNewsModel to be consistent with naming on API client.

mawadkittos commented on src/modules/app/StateType.js:
```diff
@@ -21,6 +23,10 @@ export type StateType = {
   categories: Payload<CategoriesMapModel>,
   extras: Payload<Array<ExtraModel>>,
   events: Payload<Array<EventModel>>,
+  news: Payload<Array<LocalNewsModel>>,
```
@amusameh should be localNews?

mawadkittos commented on src/modules/app/StateType.js:
```diff
@@ -21,6 +23,10 @@ export type StateType = {
   extras: Payload<Array<ExtraModel>>,
   events: Payload<Array<EventModel>>,
+  news: Payload<Array<LocalNewsModel>>,
+  newsElement: Payload<LocalNewsModel>,
```
should be localNewsElement?

mawadkittos commented on src/modules/app/reducers/index.js:
```diff
@@ -5,8 +5,12 @@ import {
   EVENTS_ENDPOINT_NAME,
+  LOCALNEWS_ENDPOINT_NAME,
+  LOCALNEWS_ELEMENT_ENDPOINT_NAME,
+  TUNEWS_ELEMENT_ENDPOINT_NAME,
```
@amusameh are we missing tunews list endpoint here?

amusameh replied:
No we are not missing it. tunews list has its own reducer

amusameh replied:
tunews list needed a separate reducer to handle the pagination. the general reducer works for the rest

mawadkittos commented on src/modules/app/route-configs/LocalNewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,57 @@
+// @flow
+
```
No need for this line

amusameh replied:
There must be a line after flow annotation 
`Expected newline after flow annotation`

mawadkittos commented on src/modules/app/route-configs/LocalNewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,57 @@
+
+  getRequiredPayloads = (payloads: AllPayloadsType): RequiredPayloadsType => ({ localNewsDetails: payloads.newsElementPayload })
+
+  getPageTitle = ({ t, payloads, cityName, location }) => null
```
@amusameh why we return null here?

mawadkittos commented on src/modules/app/route-configs/LocalNewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,57 @@
+  requiresHeader = true
+  requiresFooter = true
+
+  getLanguageChangePath = ({ location, payloads, language }) => null
```
why we return null here

amusameh replied:
Language shouldn't change for localNewsDetails route so we return `null` here. same as other routes

mawadkittos commented on src/modules/app/route-configs/NewsRouteConfig.js:
```diff
@@ -0,0 +1,77 @@
+      return null
+    }
+    const pathname = location.pathname
+    const news = payloads.news.data
```
I suggest to create method called getNewsItem and use it here and on the different places.... DRY

jafkiitos commented on src/routes/news/containers/NewsController.js:
```diff
@@ -0,0 +1,71 @@
+  handleRedirect = () => {
+    const { type, language, city, cities, redirect } = this.props
+    const currentCity: CityModel = cities && cities.find(cityElement => cityElement._code === city)
+
```
We need comments to explain the logic here

mawadkittos commented on src/modules/common/components/FailureSwitcher.js:
```diff
@@ -32,6 +34,15 @@ export class FailureSwitcher extends React.Component<PropsType> {
         return <Failure goToPath={new EventsRouteConfig().getRoutePath({ city, language })}
                         goToMessage='goTo.events'
                         errorMessage='notFound.event' />
+      case 'newsItem':
```
I suggest to move switch cases onto constants

amusameh replied:
The existed code using texts so I think we should stick to this. And since this is used in one place only I don't see an issue in having it as text

mawadkittos replied:
That's fine

mawadkittos commented on src/routes/news/actions/fetchTunews.js:
```diff
@@ -0,0 +1,28 @@
+import type { StateType } from '../../../modules/app/StateType'
+import type { StoreActionType } from '../../../modules/app/StoreActionType'
+
+const DEFAULT_LIMIT = 20
```
What do you mean by DEFAULT_LIMIT. Can you please choose more descriptive name

mawadkittos commented on src/routes/news/components/Tabs.js:
```diff
@@ -0,0 +1,49 @@
+  t: TFunction
+|}
+
+class Tabs extends React.PureComponent<PropsType> {
```
I think we don't need for Tabs component, since Tabs is a generic component also Tabs items usually should have same style, but in our case we have two special buttons with different styles, so i suggest to create two components, one for LocalNewsLink and another one for TuNewsLink.

What do you think?

mawadkittos commented on src/routes/news/containers/LocalNewsDetails.js:
```diff
@@ -0,0 +1,49 @@
+  render () {
+    const { localNewsDetails, language, city, path } = this.props
+
+    if (!localNewsDetails) {
```
localNewsDetails is an object, i think this check is not safe, since if the localNewsDetails is empty object so it will be !{} the result will be false, unless we are 100% guarantee it will return null, or undefined.

amusameh replied:
This will be always null or the data object. so I believe it's safe to use it like this.

mawadkittos replied:
That's fine

jafkiitos commented on locales/locales.json:
```diff
@@ -851,6 +851,26 @@
+      "readMore": "Read more >",
+      "currentlyNoNews": "No news is currently available.",
+      "eNewsNo": "E-news No"
+    }
```
Task for Integreat team: Add translation for the rest of the languages

klinzo replied:
This will be done by a translation agency, so it is fine to omit other languages for now. However, these changes will have to be moved to the `locales` repo, but I'll take care of that.

klinzo replied:
Do you need more locales for the native app? If so, I could do this together.

jafkiitos replied:
@klinzo The React Native app is using the same keys.

klinzo replied:
@amusameh I've created a PR in our locales repo. After it is merged, please revert the locales file here and pull them from the locales repo using git subtrees. How to do it is specified here: https://github.com/Integreat/integreat-locales

jafkiitos replied:
Thanks @klinzo. Will revert these changes.

klinzo replied:
Jfyi: On native we don't use subtrees yet, but there is a open PR which will hopefully be merged tomorrow.

klinzo commented on src/modules/common/components/LastUpdateInfo.js:
```diff
@@ -7,7 +7,8 @@ import type Moment from 'moment'
 
 const TimeStamp = styled.p`
-  padding-top: 15px;
+  margin: 0 !important;
```
All right, I understand. Rather use conditional styling here instead of wrapping it (as in `Page` component).

klinzo commented on src/modules/common/errors/ContentNotFoundError.js:
```diff
@@ -2,7 +2,7 @@
 import { ResponseError } from '@integreat-app/integreat-api-client'
 
-type NotFoundType = 'category' | 'event' | 'extra' | 'poi'
+type NotFoundType = 'category' | 'event' | 'extra' | 'poi' | 'newsItem' | 'tunewsItem'
```
For consistency I'd use `localNewsItem`.

klinzo commented on src/modules/layout/containers/LocationHeader.js:
```diff
@@ -57,12 +62,14 @@ export class LocationHeader extends React.Component<PropsType> {
 
     const isEventsActive = events ? events.length > 0 : false
     const isCategoriesEnabled = isExtrasEnabled || isEventsEnabled
+    const isNewsActive = true
```
Should be inlined.

amusameh replied:
Fixed

klinzo commented on src/modules/layout/containers/LocationHeader.js:
```diff
@@ -57,12 +62,14 @@ export class LocationHeader extends React.Component<PropsType> {
     const currentRoute = location.type
 
     const isEventsActive = events ? events.length > 0 : false
     const isCategoriesEnabled = isExtrasEnabled || isEventsEnabled
```
```suggestion
    const isCategoriesEnabled = isExtrasEnabled || isEventsEnabled || isNewsEnabled
```

klinzo commented on src/modules/layout/containers/LocationHeader.js:
```diff
@@ -90,6 +97,20 @@ export class LocationHeader extends React.Component<PropsType> {
+        <HeaderNavigationItem
+          key='news'
+          href={newsUrl}
+          selected={[LOCAL_NEWS_ROUTE, TUNEWS_LIST_ROUTE, TUNEWS_DETAILS_ROUTE, LOCAL_NEWS_DETAILS_ROUTE, LOCAL_NEWS_DETAILS_ROUTE].includes(currentRoute)}
```
```suggestion
          selected={[LOCAL_NEWS_ROUTE, TUNEWS_LIST_ROUTE, TUNEWS_DETAILS_ROUTE, LOCAL_NEWS_DETAILS_ROUTE].includes(currentRoute)}
```

amusameh replied:
Fixed

klinzo commented on src/modules/layout/containers/LocationLayout.js:
```diff
@@ -112,12 +114,14 @@ export class LocationLayout extends React.Component<PropsType, LocalStateType> {
     return <Layout asideStickyTop={this.state.asideStickyTop}
-                   header={<LocationHeader isEventsEnabled={cityModel.eventsEnabled}
-                                           isExtrasEnabled={cityModel.extrasEnabled}
+      header={<LocationHeader isEventsEnabled={cityModel && cityModel.eventsEnabled}
```
Indentation

klinzo replied:
Also, no need to check for `cityModel` as it is already handled above.

klinzo commented on src/modules/layout/containers/__tests__/__snapshots__/LocationLayout.spec.js.snap:
```diff
@@ -30,7 +30,7 @@ exports[`LocationLayout should show GeneralHeader and GeneralFooter if city is n
             "_featuredImage": undefined,
             "_hash": "2fe6283485a93932",
-            "_lastUpdate": "2016-01-07T09:36:24.000Z",
+            "_lastUpdate": "2016-01-07T08:36:24.000Z",
```
Different timezones striking again... Please revert this and the change below.

amusameh commented on src/modules/app/route-configs/CategoriesRouteConfig.js:
```diff
@@ -33,6 +34,7 @@ const categoriesRoute: Route = {
     await Promise.all([
       fetchData(createCitiesEndpoint(cmsApiBaseUrl), dispatch, state.cities),
       fetchData(createEventsEndpoint(cmsApiBaseUrl), dispatch, state.events, { city, language }),
+      fetchData(createLocalNewsEndpoint(cmsApiBaseUrl), dispatch, state.localNews, { city, language }),
```
I thought since the events are fetched here it make sense to fetch local news as well.

klinzo replied:
We have to fetch events to decide whether there are any available. If there are none, we prevent navigating to the events route. Admittedly, not a good solution though.

amusameh replied:
I see. It's not needed then, I will remove it.

klinzo commented on locales/locales.json:
```diff
@@ -851,6 +851,26 @@
+    },
+    "en": {
+      "local": "LOCAL NEWS",
+      "readMore": "Read more >",
```
Imo the arrow should not be part of the locales, what do you think?

amusameh replied:
Yeah you are right

klinzo commented on src/modules/theme/constants/helpers.js:
```diff
@@ -10,3 +10,17 @@ export default {
   `
 }
+
+export const textTruncator = (text: string, numOfWordsAllowed: number): string => {
```
This does not belong here.

amusameh replied:
I'll move it the `common` utils, What do you think?

klinzo replied:
Sounds good

klinzo commented on src/modules/theme/constants/helpers.js:
```diff
@@ -10,3 +10,17 @@ export default {
+
+export const textTruncator = (text: string, numOfWordsAllowed: number): string => {
+  const ellipsis = '...'
+  const splitTextToWords = text.split(' ')
+  if (splitTextToWords.length < numOfWordsAllowed) {
+    return text
+  }
+
+  const shortenedArrayOfText = splitTextToWords.splice(0, numOfWordsAllowed)
+  const joinShorterendArray = shortenedArrayOfText.join(' ')
+  const textTruncated = joinShorterendArray + ellipsis
+
+  return textTruncated
```
I don't really like this solution, but I am not sure if there is a better one...

amusameh replied:
Would be happy to learn a new solution. Please let me know if you have something better

klinzo replied:
I guess this is fine, can't come up with something better.

klinzo commented on src/routes/news/actions/fetchTunews.js:
```diff
@@ -0,0 +1,28 @@
+const TUNEWS_ITEMS_PER_PAGE = 20
+const DEFAULT_PAGE_NUMBER = 1
+
+export const RESET_TUNEWS = 'RESET_TUNEWS'
```
Should probably also be `TUNEWS_ENDPOINT_NAME.toUpperCase()`.

klinzo commented on src/routes/news/components/NewsElement.js:
```diff
@@ -0,0 +1,105 @@
+    return (
+      <StyledNewsElement>
+        <Link to={itemPath}>
+          {!!newsItem.title && this.renderContent(itemPath)}
```
Why do we have to do this check? It should not be necessary.

klinzo commented on src/routes/news/components/NewsElement.js:
```diff
@@ -0,0 +1,105 @@
+
+const StyledNewsElement = styled.div`
+  padding-bottom: 2px;
+  background: linear-gradient(to left, rgba(168, 168, 168, 0.2), #bebebe 51%, rgba(168, 168, 168, 0.2));
```
Where do you get these colors from?

amusameh replied:
These are the colors for the `linear gradient` for the horizontal line separator between the list items (from the designs)

amusameh replied:
We've uploaded the designs to  `Zeplin` . And that's how I got these colors.
I can check the current theme color and pick something close to the design.

[
![Annotation 2020-05-14 124042](https://user-images.githubusercontent.com/28042978/81919207-343f5480-95e0-11ea-8916-98b0c2719f84.jpg)
](url)

klinzo replied:
Okay, sorry for our bad design and communication here then, I would appreciate to use a similar color from our theme. (with exception of the two colors for TüNews).

amusameh replied:
Just to double check. you mean by `two colors for TüNews` the light and dark blueish colors.
 
But the  `content` and `title` color will be `black` as the rest of the website. right?

amusameh replied:
@klinzo also should I keep the color gradient for the items horizontal separator or replace it with a solid border with a color from the theme maybe `textDisabledColor`

klinzo replied:
Didn't know about the light blue. So if you need the grey tünews color too, you can also add it to the theme. If you'd say it looks good enough to use one of our defined colors, use those.

klinzo replied:
Is there a reason why we should not keep it? Imo it is fine :+1:

amusameh replied:
@klinzo I think we should keep it as well. Anything still needs to be updated for this comment?

klinzo commented on src/routes/news/components/NewsElement.js:
```diff
@@ -0,0 +1,105 @@
+
+const Link = styled(CleanLink)`
+  display: flex;
+  background-color: white;
```
Please don't use colors like this but instead use those defined in the theme.

amusameh replied:
What if the color is not defined in the theme and used once or twice?

klinzo replied:
Fixed as you are using background color now, right?

amusameh replied:
Yeah

klinzo commented on src/routes/news/components/NewsElement.js:
```diff
@@ -0,0 +1,105 @@
+      <>
+        <Description>
+          <Title>{newsItem.title}</Title>
+          <Body>{newsItem && textTruncator(newsItem.message, TRUNCATE_LETTERS_COUNT)}</Body>
```
No need to check for `newsItem`.

klinzo commented on src/routes/news/components/NewsElement.js:
```diff
@@ -0,0 +1,105 @@
+  renderContent (itemPath: string): React.Node {
+    const { newsItem, language, t } = this.props
+    return (
+      <>
```
Should not necessary.

klinzo commented:
Please also check for other occurrences of my requested changes and adjust if necessary. Especially for
* updated timestamps in snapshots
* usage of colors
* usage of `!important`
* unnecessary checks for null/undefined

amusameh commented on src/modules/app/route-configs/TunewsListRouteConfig.js:
```diff
@@ -0,0 +1,72 @@
+      fetchData(createLocalNewsEndpoint(cmsApiBaseUrl), dispatch, state.localNews, { city, language })
+    ]
+
+    if (!state.tunews.data.length) {
```
It's not needed. I fixed that

mawadkittos commented on src/modules/app/StateType.js:
```diff
@@ -21,6 +23,10 @@ export type StateType = {
   extras: Payload<Array<ExtraModel>>,
   events: Payload<Array<EventModel>>,
+  localNews: Payload<Array<LocalNewsModel>>,
+  localNewsElement: Payload<LocalNewsModel>,
```
I think it make sense here, since we have separate endpoint for each of them, by this we have control on the response type, for example if the endpoint for localNews return different fields from what localNewsElement endpoint return.
Does that make sense?

amusameh replied:
Since we have separate endpoints I think it make sense to be like this. What are you suggesting here?

klinzo replied:
Yeah I also think so. Would be too much overhead changing it to one element. So it's fine.

amusameh commented on src/modules/theme/constants/colors.js:
```diff
@@ -8,7 +8,9 @@ export type ColorsType = {
   textDecorationColor: string,
   textDisabledColor: string,
-  backgroundColor: string
+  headlineTextColor: string,
```
This was to exactly match the designs color. I could change it to `black` if you prefer that

jafkiitos replied:
@klinzo @amusameh in such case, I guess they should be taken out of the theme file and used right in the Tunews file, as this is not a generic theme color and should not exist here.

klinzo replied:
With right naming I don't have a problem leaving it in the theme, but I also don't mind moving it to a tunews constant file or similar. The decision is up to you.

amusameh replied:
This should be fixed

amusameh commented on src/modules/app/route-configs/LocalNewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,70 @@
+
+  getMetaDescription = () => null
+
+  getFeedbackTargetInformation = ({ payloads, location }) => {
```
@klinzo Did you agree on this?

klinzo replied:
Just return null, feedback is not available.

amusameh replied:
Done

klinzo commented on src/modules/app/route-configs/__tests__/__snapshots__/index.spec.js.snap:
```diff
@@ -26,7 +26,15 @@ Object {
+    "thunk": [Function],
+  },
   "MAIN_DISCLAIMER": "/disclaimer",
+  "NEWS": Object {
```
Should be `LOCAL_NEWS`

klinzo commented on src/modules/layout/containers/LocationHeader.js:
```diff
@@ -30,7 +33,9 @@ type PropsType = {|
   t: TFunction,
   cityName: string,
   isEventsEnabled: boolean,
+  isNewsEnabled: boolean,
```
I think it would be more intuitive if we would pass `isLocalNewsEnabled` instead of the combined `isNewsEnabled`. Otherwise one does not really now imo if its just local news or both.

amusameh replied:
The `isNewsEnabled` represent both `localNews` and `tunews`. I mean it's for showing and hiding  the navigation `news` link. calling it `isLocalNewsEnabled` wouldn't be accurate I think.

klinzo replied:
I didn't mean to just change the name of it, but rather just pass `isLocalNewsEnabled` and `isTunewsEnabled` separately instead of doing `isNewsEnabled = isLocalNewsEnabled || isTuNewsEnabled` in the `LocationHeaderContainer`. Makes the code here way clearer imo.

amusameh commented on src/modules/app/route-configs/LocalNewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,70 @@
+import fetchData from '../fetchData'
+import { cmsApiBaseUrl } from '../constants/urls'
+
+type LocalNewsDetailsType = {| city: string, language: string, id: string | number |}
```
Fixed

amusameh commented on src/modules/app/route-configs/TunewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,69 @@
+import fetchData from '../fetchData'
+import { cmsApiBaseUrl, tunewsApiBaseUrl } from '../constants/urls'
+
+type TunewsDetailsRouteParamsType = {| city: string, language: string, id: string | number |}
```
Fixed. Changed it to number. as the id is number in the api response

klinzo commented on src/routes/categories/containers/__tests__/__snapshots__/CategoriesPage.spec.js.snap:
```diff
@@ -90,7 +90,7 @@ exports[`CategoriesPage should match snapshot and render a Page if page has no c
     content="exampleContent0"
     language="en"
-    lastUpdate={"2016-01-07T09:36:24.000Z"}
+    lastUpdate={"2016-01-07T08:36:24.000Z"}
```
We may have to set a timezone for jest: https://stackoverflow.com/questions/56261381/how-do-i-set-a-timezone-in-my-jest-config

amusameh replied:
Thanks. should be Fixed. I used the config below to change the tz and pushed the updated snapshots without the code changes to the jest configs as I'm not sure it should be in this pr or not
```js
module.exports = async () => {
  process.env.TZ = 'Europe/Berlin'
}
```

klinzo replied:
@maxammann what do you think? Should this be pushed? I'm inclined to say yes because we had this problem already some times.

amusameh replied:
@maxammann @klinzo Please let me know when you decide on this

amusameh commented on src/modules/app/route-configs/TunewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,69 @@
+
+  getMetaDescription = () => null
+
+  getFeedbackTargetInformation = ({ payloads }) => {
```
Fixed

amusameh commented on src/modules/app/route-configs/TunewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,69 @@
+    return `${tunewsElementDetails.title} - ${cityName}`
+  }
+
+  getRoutePath = ({ city, language, id }): string => `/${city}/${language}/news/local/${id}`
```
Fixed

klinzo commented on src/routes/news/actions/fetchMoreTunews.js:
```diff
@@ -0,0 +1,23 @@
+// @flow
```
Should not be part of the route, but probably rather in module > app > actions

amusameh replied:
In the routes `readme` :
`This package contains all the routes. Routes can contain containers, components, hocs, actions and reducers`

I think as this is specific to tunews it should stay here. What do you think?

maxammann replied:
It depends whether this action is used from multiple routes. If it is only used within `routes/news/` then this is fine.

klinzo commented on src/routes/news/actions/finishFetchMoreAction.js:
```diff
@@ -0,0 +1,14 @@
+// @flow
```
Should not be part of the route, but probably rather in module > app > actions

maxammann replied:
It depends whether this action is used from multiple routes. If it is only used within routes/news/ then this is fine.

amusameh replied:
Yeah I agree. It could be used on other routes. I will move it to module > app > actions

klinzo commented on src/routes/news/actions/startFetchMoreAction.js:
```diff
@@ -0,0 +1,16 @@
+// @flow
```
Should not be part of the route, but probably rather in module > app > actions

maxammann replied:
It depends whether this action is used from multiple routes. If it is only used within routes/news/ then this is fine.

maxammann commented on src/modules/app/route-configs/LocalNewsDetailsRouteConfig.js:
```diff
@@ -0,0 +1,67 @@
+  route = localNewsDetailsRoute
+  isLocationLayoutRoute = true
+  requiresHeader = true
+  requiresFooter = true
```
@klinzo do you remember what these options were for? Are they correct?

klinzo replied:
Yes, this defines whether a header and footer are rendered, e.g. for the LandingPage we don't want a header.
It is correct.

maxammann commented on src/routes/news/components/TunewsDetailsFooter.js:
```diff
@@ -0,0 +1,56 @@
+
+const Footer = styled.footer`
+  display: flex;
+  width: 96%;
```
This width seems a little bit weird. Was this applied in order to fix some weird behavior?

amusameh replied:
Yeah. I changed that.

maxammann commented on src/routes/news/components/TunewsDetailsFooter.js:
```diff
@@ -0,0 +1,56 @@
+        <StyledContainer>{t('eNewsNo')}: {eNewsNumber}</StyledContainer>
+        <StyledContainer>
+          <StyledLink href='http://www.tunews.de' target='_blank'>
+            Tunews International
```
```suggestion
            tünews INTERNATIONAL
```

@svenseeberg what is the correct name of tünews?

amusameh replied:
Fixed as the suggestion

maxammann commented on src/routes/news/components/TunewsDetailsFooter.js:
```diff
@@ -0,0 +1,56 @@
+      <Footer>
+        <StyledContainer>{t('eNewsNo')}: {eNewsNumber}</StyledContainer>
+        <StyledContainer>
+          <StyledLink href='http://www.tunews.de' target='_blank'>
```
@svenseeberg can we maybe link to https://tunewsinternational.com/ here?

maxammann replied:
This should be changed to https://tunewsinternational.com

amusameh replied:
Fixed!

maxammann commented on src/routes/news/components/__tests__/LocalNewsList.spec.js:
```diff
@@ -0,0 +1,52 @@
+
+  const items = [localNews1, localNews2]
+
+  it('should render and match snapshot', () => {
```
It would be very good if this tests would assert fro actual behaviour of the component. Like that it displays some items.

amusameh replied:
Should be fixed. I added a couple more test cases.

maxammann replied:
@amusameh I think you can remove this snapshot tests as it doesn't really test anything

maxammann commented on src/routes/news/components/__tests__/NewsElement.spec.js:
```diff
@@ -0,0 +1,31 @@
+  it('should render and match snapshot', () => {
+    expect(shallow(
+      <NewsElement id={id} type={LOCAL_NEWS} title={title} content={message} timestamp={timestamp} language={language} t={t} path={path} />
+    )).toMatchSnapshot()
```
It would be very good if this tests would assert fro actual behavior of the component. Like that it displays some items.

amusameh replied:
Should be fixed

maxammann replied:
@amusameh I think you can remove this snapshot tests as it doesn't really test anything

maxammann commented on src/routes/news/components/__tests__/NewsTabs.spec.js:
```diff
@@ -0,0 +1,78 @@
+        t={t}>
+          <div>dummy child</div>
+      </NewsTabs>
+    )).toMatchSnapshot()
```
It would be very good if this tests would assert fro actual behavior of the component. Like that it displays some items.

maxammann commented on src/routes/news/components/__tests__/Tab.spec.js:
```diff
@@ -0,0 +1,24 @@
+        destination={destination}
+        t={t}
+      />
+    )).toMatchSnapshot()
```
It would be very good if this tests would assert fro actual behavior of the component. Like that it displays some items.

maxammann commented on src/routes/news/components/__tests__/TunewsDetailsFooter.spec.js:
```diff
@@ -0,0 +1,23 @@
+        language={language}
+        t={t}
+      />
+    )).toMatchSnapshot()
```
It would be very good if this tests would assert fro actual behavior of the component. Like that it displays some items.

maxammann commented on src/routes/news/containers/LocalNewsDetailsPage.js:
```diff
@@ -0,0 +1,53 @@
+  }
+)
+
+export default connect<*, *, *, *, *, *>(mapStateTypeToProps)(LocalNewsDetailsPage)
```
Proper flow typing is missing here. There should be some examples in the RN app or webapp which show how to do this.

amusameh replied:
It's used like this in webapp everywhere. I'll take a look at the RN app.

amusameh replied:
The RN is using `connect<PropsType, OwnPropsType, _, _, _, _>(mapStateToProps, mapDispatchToProps)`

Since I'm not using ownProps I'm not sure if I should add it. Please correct me if I'm wrong. I will use this `connect<PropsType, *, *, *, *, *>`

@maxammann thoughts please?

amusameh replied:
Should be fixed

maxammann commented on src/routes/news/containers/LocalNewsPage.js:
```diff
@@ -0,0 +1,66 @@
+}
+
+export default compose(
+  connect<*, *, *, *, *, *>(mapStateTypeToProps),
```
Proper flow typing is missing here. There should be some examples in the RN app or webapp which show how to do this.

amusameh replied:
Should be fixed

maxammann commented on src/routes/news/containers/NewsController.js:
```diff
@@ -0,0 +1,74 @@
+  redirect: action => dispatch(redirect(action))
+})
+
+export default connect<*, *, *, *, *, *>(mapStateTypeToProps, mapDispatchToProps)(NewsController)
```
Proper flow typing is missing here. There should be some examples in the RN app or webapp which show how to do this.

amusameh replied:
Should be fixed

maxammann commented on src/routes/news/containers/TunewsDetailsPage.js:
```diff
@@ -0,0 +1,105 @@
+})
+
+export default compose(
+  connect<*, *, *, *, *, *>(mapStateToProps),
```
Proper flow typing is missing here. There should be some examples in the RN app or webapp which show how to do this.

amusameh replied:
Should be fixed

maxammann commented on src/routes/news/containers/__tests__/LocalNewsPage.spec.js:
```diff
@@ -0,0 +1,112 @@
+    })
+  })
+
+  moment.tz.setDefault()
```
This call leaks the scope of this test. Why is this needed?

klinzo replied:
I guess this is needed to prevent failing of tests if in different timezones, right? See this comment: https://github.com/Integreat/integreat-webapp/pull/333#discussion_r424237322

amusameh replied:
This is not needed anymore as per the above comment discussion.
It should be fixed.

maxammann commented on src/routes/news/containers/__tests__/NewsController.spec.js:
```diff
@@ -0,0 +1,95 @@
+    })
+  })
+
+  moment.tz.setDefault()
```
This call leaks the scope of this test. Why is this needed?

amusameh replied:
Fixed.

maxammann commented on src/routes/news/containers/__tests__/TunewsDetailsPage.spec.js:
```diff
@@ -0,0 +1,110 @@
+    })
+  })
+
+  moment.tz.setDefault()
```
This call leaks the scope of this test. Why is this needed?

amusameh replied:
Fixed.

maxammann commented on src/routes/news/containers/__tests__/LocalNewsDetailsPage.spec.js:
```diff
@@ -0,0 +1,76 @@
+    })
+  })
+
+  moment.tz.setDefault()
```
This call leaks the scope of this test. Why is this needed?

amusameh replied:
Fixed.

maxammann commented on src/routes/news/containers/__tests__/TunewsPage.spec.js:
```diff
@@ -0,0 +1,125 @@
+    })
+  })
+
+  moment.tz.setDefault()
```
This call leaks the scope of this test. Why is this needed?

amusameh replied:
Fixed.

klinzo commented on src/routes/news/fetchMoreData.js:
```diff
@@ -0,0 +1,40 @@
+// @flow
```
I also think this should be moved to `app`. Has to be discuessed with @maxammann though.

klinzo commented on src/routes/news/fetchMoreData.js:
```diff
@@ -0,0 +1,40 @@
+import finishFetchMoreAction from './actions/finishFetchMoreAction'
+import type { StoreActionType } from '../../modules/app/StoreActionType'
+
+async function fetchData<P, T> (
```
Is this different from the actual `fetchData`? Or is the only difference what actions are dispatched? If so, we could maybe also just use a flag for this and use the actual `fetchData`.

amusameh replied:
I thought about passing the actions as a parameter instead of making it global but I wasn't sure about it that's why I created a new function. I think it would be cleaner than using a flag if you ever needed to dispatch other actions
What do you think? @klinzo  @maxammann

klinzo replied:
We could pass two functions as optional params which map endpoint and payload to start/finish action. But I'm also not sure. @maxammann what do you think?

klinzo replied:
Still relevant

amusameh replied:
tbh. I'm not sure what to do here? Have you talked to @maxammann about this?

klinzo replied:
Yeah sure, sorry. We have to wait for an answer here. My comment was just meant as a hint for myself and @maxammann

klinzo commented on src/routes/news/reducers/index.js:
```diff
@@ -0,0 +1,54 @@
+        data: [...action.payload.data],
+        requestUrl: action.payload.requestUrl,
+        isFetching: state.isFetching,
+        hasMore: action.payload.data.length !== 0,
```
Is this correct? Seems weird to me.

amusameh replied:
Yeah. If the last request(while scrolling down) is an empty array then `hasMore` is set to `false` to stop sending more requests

amusameh commented on src/routes/news/components/Tab.js:
```diff
@@ -0,0 +1,69 @@
+  color: ${({ theme }) => (theme.colors.backgroundColor)};
+  object-fit: contain;
+  background-color: ${({ active, theme }) =>
+    active ? theme.colors.themeColor : 'rgba(111, 111, 110, 0.4)'};
```
This button color. I guess I could replace it with this theme color `textDisabledColor`
![Annotation 2020-05-14 171434](https://user-images.githubusercontent.com/28042978/81945176-69f73400-9606-11ea-8d00-dcbade1bdb99.jpg)

klinzo replied:
Thanks!

klinzo commented on src/routes/news/components/LocalNewsList.js:
```diff
@@ -0,0 +1,46 @@
+const StyledList = styled.div`
+  position: relative;
+  padding-top: 1px;
+  background: linear-gradient(to left, rgba(168, 168, 168, 0.2), #bebebe 51%, rgba(168, 168, 168, 0.2));
```
Is it possible to use hex colors instead of rgba here?

Edit: https://stackoverflow.com/questions/15852122/hex-transparency-in-colors

klinzo replied:
Also, why 51% instead of just 50%?

amusameh replied:
ASFAIK rgba is more supported than `8 character hex notation` so I prefer to use it. 
As for the 51% I copied the style from the design but I probably should be 50% I'll change it.

amusameh replied:
@klinzo Is there anything still relevant here?

klinzo replied:
I guess not, nope.

klinzo commented on src/routes/news/components/NewsTabs.js:
```diff
@@ -0,0 +1,60 @@
+              <Tab
+                active={type === LOCAL_NEWS}
+                type={LOCAL_NEWS}
+                destination={`/${city}/${language}/news/local`}
```
We should probably use the corresponding method of the route config instead of hard coding the path here.

klinzo commented on src/modules/theme/constants/colors.js:
```diff
@@ -19,7 +21,9 @@ export const brightColors: ColorsType = {
-  backgroundColor: '#ffffff'
+  backgroundColor: '#ffffff',
+  tunewsThemeColor: '#007aa8',
+  tunewsThemeColorLight: 'rgba(0, 122, 168, 0.4)'
```
If possible this should be hex.

amusameh replied:
ASFAIK rgba is more supported than 8 character hex notation so I prefer to use it.
It's an easy switch if you think browser support won't be an issue

klinzo replied:
No, that's fine.

klinzo commented on src/routes/news/components/TunewsList.js:
```diff
@@ -0,0 +1,59 @@
+
+class TunewsList extends React.PureComponent<PropsType> {
+  loadItems = async (page: number) => {
+    this.props.fetchMoreTunews(page + 1, LIMIT)
```
I think we should either await this or loadItems shouldn't be async.

amusameh replied:
Fixed. deleted the `async`

klinzo commented on src/routes/news/containers/NewsController.js:
```diff
@@ -0,0 +1,74 @@
+    this.handleRedirect()
+  }
+
+  handleRedirect = () => {
```
Did you remove it completely? If so, why?

amusameh replied:
Yeah I removed it. I think it make more sense to show not found error instead of redirecting the user to another page. (I'll make some changes and push it)
before => if the user enters tunews link directly and tunews are disabled for this city it will redirect instead of showing not found error
@klinzo What do you think?

klinzo commented:
I think a bit of space beneath the list would be good: 
![image](https://user-images.githubusercontent.com/29095487/82138969-e9fae500-9824-11ea-9c0c-4a3771d93c72.png)

klinzo commented:
Also, if you navigate from the tünews list to another page (e.g. tünews detail) and then navigate back, everything is fetched again. I think we should not fetch anything if there already is data in the state.

klinzo commented:
Where do you test local news, disabled local news and disabled tünews? (Which instance, I'm not sure where we have according settings).

amusameh commented on src/routes/news/components/NewsElement.js:
```diff
@@ -0,0 +1,108 @@
+  timestamp: Moment,
+  language: string,
+  path: string,
+  type: string,
```
Fixed

amusameh commented:
> I think a bit of space beneath the list would be good:
> ![image](https://user-images.githubusercontent.com/29095487/82138969-e9fae500-9824-11ea-9c0c-4a3771d93c72.png)

Fixed. I added `40px` margin to match the top space between the list and the tabs

amusameh commented:
> Where do you test local news, disabled local news and disabled tünews? (Which instance, I'm not sure where we have according settings).

This city `testumgebung` has both localNews and tunews enabled. It used to have some local news test data but it's deleted now (maybe yesturday) [check this link](https://cms-test.integreat-app.de/testumgebung/de/wp-json/extensions/v3/fcm?channel=news) 
This city `konstanz` has only localNews enabled
There is not city where tunews only enabled. Maybe you can add one and test that

amusameh commented:
> Also, if you navigate from the tünews list to another page (e.g. tünews detail) and then navigate back, everything is fetched again. I think we should not fetch anything if there already is data in the state.

Should be fixed now

klinzo commented on src/routes/news/components/Tab.js:
```diff
@@ -13,7 +13,7 @@ const NewsLink = ({ active, ...props }) => <Link {...props} />
   display: flex;
   width: 160px;
-  height: 47px;
+  height: 49px;
```
Just make it 50 :smile:

amusameh replied:
done

klinzo commented on src/routes/news/constants.js:
```diff
@@ -1,4 +1,6 @@
 // @flow
 
+export type NewselementType = 'LOCAL_NEWS' | 'TU_NEWS'
```
```suggestion
export type NewsType = 'LOCAL_NEWS' | 'TU_NEWS'
```
should be enough. If you don't think so, change it to `NewsElementType` with a upper E.

amusameh replied:
Fixed

klinzo commented on src/routes/news/containers/TunewsPage.js:
```diff
@@ -48,6 +54,17 @@ export class TunewsPage extends React.PureComponent<PropsType> {
   render () {
     const { tunews, language, city, t, fetchMoreTunews, hasMore, isFetchingFirstTime, isFetching, cities } = this.props
+
+    if (this.props.areCitiesFetching) {
```
AFAICS this is missing in the props. Is this needed?

amusameh replied:
Fixed! Yes it's needed or there will be an error because the cities are null while fetching

klinzo commented on src/routes/news/containers/TunewsPage.js:
```diff
@@ -48,6 +54,17 @@ export class TunewsPage extends React.PureComponent<PropsType> {
+
+    const currentCity: CityModel = cities && cities.find(cityElement => cityElement.code === city)
+
+    if (!currentCity.tunewsEnabled) {
```
As discussed: A `ContentNotFoundError` should be displayed here.

amusameh replied:
Fixed!

klinzo commented:
> > Where do you test local news, disabled local news and disabled tünews? (Which instance, I'm not sure where we have according settings).
> 
> This city `testumgebung` has both localNews and tunews enabled. It used to have some local news test data but it's deleted now (maybe yesturday) [check this link](https://cms-test.integreat-app.de/testumgebung/de/wp-json/extensions/v3/fcm?channel=news)
> This city `konstanz` has only localNews enabled
> There is not city where tunews only enabled. Maybe you can add one and test that

Thanks!

klinzo commented:
@amusameh How did you solve the issue with the timezones in the snapshot tests?

amusameh commented:
@klinzo 
> @amusameh How did you solve the issue with the timezones in the snapshot tests?
I edit the jest global setup to use Germany time

1. new file => `src/global-setup.js`
```js
// @flow

module.exports = async () => {
  process.env.TZ = 'Europe/Berlin'
}
```
2. Added this line to the end of  `jest.config.json`
`"globalSetup": "<rootDir>/global-setup.js"`

klinzo commented:
Thanks!

klinzo commented on src/routes/news/components/TunewsDetailsFooter.js:
```diff
@@ -47,7 +47,7 @@ class TunewsDetailsFooter extends React.PureComponent<PropsType> {
     return (
       <Footer>
-        <StyledContainer>{t('eNewsNo')}: {eNewsNo}</StyledContainer>
+        <StyledContainer>E-news No: {eNewsNo}</StyledContainer>
```
Ah okay I guess we were not quite clear: It should be removed entirely, sorry for the confusion. So this should be:
```suggestion
        <StyledContainer>{eNewsNo}</StyledContainer>
```

klinzo commented:
TODO before merging: Pull locales via subtree
@amusameh should I do that?

amusameh commented:
> TODO before merging: Pull locales via subtree
> @amusameh should I do that?

Sorry i just saw this. Apparently @maxammann has done that already, thanks : )

Taggotty commented on src/modules/common/utils/textTruncator.js:
```diff
@@ -0,0 +1,17 @@
+
+const textTruncator = (text: string, numOfWordsAllowed: number): string => {
+  const ellipsis = '...'
+  const splitTextToWords = text.split(' ')
```
Better naming would be just `words`

amusameh replied:
I agree. Should I create a new PR for this since the current one is merged?

Taggotty commented on src/modules/common/components/LastUpdateInfo.js:
```diff
@@ -16,18 +16,20 @@ const TimeStamp = styled.p`
-  language: string
+  language: string,
+  withText: boolean,
+  className?: string
```
Why do you need the className?

amusameh replied:
Good morning @Taggotty

I passed the `className` to allow me tweak and customize the component style if needed using styled components
example:
```js
const CustomLastUpdateInfo = styled(LastUpdateInfo)`
  padding: 0 10px;
  color: ${({ theme }) => (theme.colors.backgroundColor)};

  &&& {
    margin: 0
  }
`
```

Taggotty commented on src/routes/news/containers/LocalNewsPage.js:
```diff
@@ -0,0 +1,85 @@
+    const currentCity: CityModel = cities && cities.find(cityElement => cityElement.code === city)
+
+    if (!currentCity.pushNotificationsEnabled) {
+      const type = currentCity.tunewsEnabled ? 'tunewsItem' : 'category'
```
shouldn't it be?
```suggestion
      const type = currentCity.tunewsEnabled ? 'tunewsItem' : 'localNewsItem'
```

klinzo replied:
No, it shouldn't. In fact, imo it should even always be `category`. If a city disables the feature, the user shouldn't notice that the feature exists at all and rather be displayed a general not found error, which is in our case a category not found.

amusameh replied:
This is meant to redirect the user to `category` route if both local news "pushNotificationsEnabled" and tunews were disabled.

amusameh commented on src/modules/app/containers/__tests__/Switcher.spec.js:
```diff
@@ -197,7 +201,8 @@ describe('Switcher', () => {
     return (
       <Switcher viewportSmall={false} location={location} citiesPayload={citiesPayload}
                 categoriesPayload={categoriesPayload} eventsPayload={eventsPayload} extrasPayload={extrasPayload}
-                poisPayload={poisPayload} disclaimerPayload={disclaimerPayload} languages={languages} t={t}
```
This line isn't deleted. I just added the new changes before it and split it into two lines so it seems like it was deleted
